File: mobile/modules/vibe-chat-native/android/src/main/java/expo/modules/vibechatnative/SecureKeyStore.kt
package expo.modules.vibechatnative

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey

/// Thread-safe encrypted storage wrapper for sensitive values (private keys, auth tokens).
/// Uses AndroidKeyStore-backed MasterKey with AES256-GCM encryption.
internal object SecureKeyStore {
  private const val PREFS_NAME = "vibe_secure_keys"
  const val SENTINEL = "__SECURE__"

  val sensitiveKeys = setOf("privateKeyPem", "privateKey", "authToken", "token")

  @Volatile
  private var cachedPrefs: SharedPreferences? = null

  private fun getPrefs(context: Context): SharedPreferences {
    cachedPrefs?.let { return it }
    synchronized(this) {
      cachedPrefs?.let { return it }
      val masterKey = MasterKey.Builder(context.applicationContext)
        .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
        .build()
      val prefs = EncryptedSharedPreferences.create(
        context.applicationContext,
        PREFS_NAME,
        masterKey,
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
      )
      cachedPrefs = prefs
      return prefs
    }
  }

  fun storeSecret(context: Context, key: String, value: String) {
    getPrefs(context).edit().putString(key, value).apply()
  }

  fun retrieveSecret(context: Context, key: String): String? {
    return getPrefs(context).getString(key, null)
  }

  fun deleteSecret(context: Context, key: String) {
    getPrefs(context).edit().remove(key).apply()
  }

  /// Migrate secrets from a plaintext config map into encrypted storage.
  /// Returns a sanitized copy with secrets replaced by sentinels.
  fun migrateAndSanitize(context: Context, config: Map<String, Any?>): Map<String, Any?> {
    val sanitized = config.toMutableMap()
    for (key in sensitiveKeys) {
      val value = config[key] as? String
      if (!value.isNullOrBlank() && value != SENTINEL) {
        storeSecret(context, key, value)
        sanitized[key] = SENTINEL
      }
    }
    return sanitized
  }

  /// Reassemble a config by replacing sentinels with values from secure storage.
  fun reassemble(context: Context, config: Map<String, Any?>): Map<String, Any?> {
    val assembled = config.toMutableMap()
    for (key in sensitiveKeys) {
      val stored = config[key] as? String
      if (stored == SENTINEL) {
        val secret = retrieveSecret(context, key)
        if (secret != null) {
          assembled[key] = secret
        }
      }
    }
    return assembled
  }
}

File: mobile/modules/vibe-chat-native/ios/SecureKeyStore.swift
import Foundation
import Security

/// Thread-safe Keychain wrapper for storing sensitive values (private keys, auth tokens).
/// Uses kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly so secrets survive reboots
/// but are NOT included in device backups or migrations.
final class SecureKeyStore {
  static let shared = SecureKeyStore()

  private let service = "vibe.chat.engine.securekeys"
  private let queue = DispatchQueue(label: "vibe.securekeystore")

  private init() {}

  func storeSecret(key: String, value: String) -> Bool {
    queue.sync {
      guard let data = value.data(using: .utf8) else { return false }
      deleteSecretLocked(key: key)
      let query: [String: Any] = [
        kSecClass as String: kSecClassGenericPassword,
        kSecAttrService as String: service,
        kSecAttrAccount as String: key,
        kSecValueData as String: data,
        kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
      ]
      let status = SecItemAdd(query as CFDictionary, nil)
      return status == errSecSuccess
    }
  }

  func retrieveSecret(key: String) -> String? {
    queue.sync {
      let query: [String: Any] = [
        kSecClass as String: kSecClassGenericPassword,
        kSecAttrService as String: service,
        kSecAttrAccount as String: key,
        kSecReturnData as String: true,
        kSecMatchLimit as String: kSecMatchLimitOne,
      ]
      var result: AnyObject?
      let status = SecItemCopyMatching(query as CFDictionary, &result)
      guard status == errSecSuccess, let data = result as? Data else { return nil }
      return String(data: data, encoding: .utf8)
    }
  }

  func deleteSecret(key: String) {
    queue.sync {
      deleteSecretLocked(key: key)
    }
  }

  private func deleteSecretLocked(key: String) {
    let query: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: service,
      kSecAttrAccount as String: key,
    ]
    SecItemDelete(query as CFDictionary)
  }

  /// Migrate secrets from a plaintext config dictionary into the Keychain.
  /// Returns a sanitized copy of the config with secrets replaced by sentinels.
  func migrateAndSanitize(_ config: [String: Any]) -> [String: Any] {
    var sanitized = config
    for key in Self.sensitiveKeys {
      if let value = config[key] as? String, !value.isEmpty, value != Self.sentinel {
        _ = storeSecret(key: key, value: value)
        sanitized[key] = Self.sentinel
      }
    }
    return sanitized
  }

  /// Reassemble a config by replacing sentinels with values from secure storage.
  func reassemble(_ config: [String: Any]) -> [String: Any] {
    var assembled = config
    for key in Self.sensitiveKeys {
      if let stored = config[key] as? String, stored == Self.sentinel {
        if let secret = retrieveSecret(key: key) {
          assembled[key] = secret
        }
      }
    }
    return assembled
  }

  static let sentinel = "__SECURE__"
  static let sensitiveKeys = ["privateKeyPem", "privateKey", "authToken", "token"]
}
